src/async_fuse/memfs/cache/memory_cache.rs

```diff
@@ -0,0 +1,439 @@
+                } else {
+                    // If a fetching from backend still fails, create a new block filled with zeros,
+                    // then write it to cache and backend as stored.
+                    let zeroed_block = Block::new(self.block_size);
+
+                    // It's ok to call `self.store()` directly, because we are going to store a whole block.
+                    self.store(
+                        ino,
+                        block_id,
+                        IoBlock::new(zeroed_block.clone(), 0, self.block_size),
+                    )
+                    .await;
+
+                    zeroed_block
+                };
+
+                Some(result)
+            }
+            Some(block) => {
+                self.policy.touch(&BlockCoordinate(ino, block_id));
+                Some(block)
+            }
+        }
+    }
+
+    async fn store(&self, ino: INum, block_id: usize, block: IoBlock) {
+        let start_offset = block.offset();
+        let end_offset = block.end();
+
+        // If the writing block is the whole block, then there is no need to fetch a block from cache or backend,
+        // as the block in storage will be overwritten directly.
+        // TODO: Mark it as dirty
+        if start_offset == 0 && end_offset == self.block_size {
+            self.write_and_evict(ino, block_id, block.block().clone())
+                .await;
+            self.backend.store(ino, block_id, block).await;
+            return;
+        }
+
+        let inserted = self.merge_block_in_place(ino, block_id, &block);
+        if inserted {
+            return;
+        }
+
+        let mut to_be_inserted = {
+            let block_from_backend = self.backend.load(ino, block_id).await;
+            block_from_backend.unwrap_or_else(|| {
+                // Create a new block for write, despite the offset is larger than file size.
+                Block::new(self.block_size)
+            })
+        };
+
+        merge_two_blocks(&mut to_be_inserted, &block);
+
+        self.write_and_evict(ino, block_id, to_be_inserted).await;
+
+        // Write through
+        // TODO: Spawn it after load.
+        self.backend.store(ino, block_id, block).await;
+    }
+
+    async fn remove(&self, ino: INum) {
+        self.map.remove(&ino);
+        self.backend.remove(ino).await;
+    }
+
+    async fn invalidate(&self, ino: INum) {
+        self.map.remove(&ino);
+        self.backend.invalidate(ino).await;
+    }
+
+    async fn flush(&self, ino: INum) {
+        self.backend.flush(ino).await;
+    }
+
+    async fn flush_all(&self) {
+        // The `InMemoryCache` cannot iterate its `HashMap`, thus it cannot `flush_all` on all the file cache.
+        // As it runs with a "write through" policy, the backend takes the responsibility to do the actual `flush_all`.
+        self.backend.flush_all().await;
+    }
+}
+
+#[cfg(test)]
+#[allow(clippy::unwrap_used)]
+mod tests {
+    use std::sync::Arc;
```

Comment:
Please make a large test function into small ones. It's easier to debug, and has better performance on power servers.